src/workerd/api/hibernatable-web-socket.c++

```diff
@@ -12,24 +12,47 @@ namespace workerd::api {
 
-HibernatableWebSocketEvent::ItemsForRelease HibernatableWebSocketEvent::prepareForRelease(
-    jsg::Lock &lock) {
+Worker::Actor::HibernationManager& HibernatableWebSocketEvent::getHibernationManager(
+    jsg::Lock& lock) {
   auto& actor = KJ_REQUIRE_NONNULL(IoContext::current().getActor());
-  auto& manager = kj::downcast<HibernationManagerImpl>(
-      KJ_REQUIRE_NONNULL(actor.getHibernationManager()));
-  auto& hibernatableWebSocket = KJ_REQUIRE_NONNULL(manager.webSocketForEventHandler);
+  return KJ_REQUIRE_NONNULL(actor.getHibernationManager());
+}
+
+HibernatableWebSocketEvent::ItemsForRelease HibernatableWebSocketEvent::prepareForRelease(
+    jsg::Lock &lock, kj::StringPtr websocketId) {
+  auto& manager = kj::downcast<HibernationManagerImpl>(getHibernationManager(lock));
+  auto& hibernatableWebSocket = *KJ_REQUIRE_NONNULL(
+      manager.webSocketsForEventHandler.find(websocketId));
+
+  // Note that we don't call `getWebSocket()` to get this, since we would lose our reference to
+  // the HibernatableWebSocket (it removes it from `webSocketsForEventHandler`).
+  auto websocketRef = hibernatableWebSocket.getActiveOrUnhibernate(lock);
+  auto ownedWebSocket = kj::mv(KJ_REQUIRE_NONNULL(hibernatableWebSocket.ws));
+
+  // Now that we've obtained the websocket for the event, let's free up the slots we had allocated.
+  manager.webSocketsForEventHandler.erase(websocketId);
 
-  // `getWebSocket()` requires `HibernatableWebSocket::ws` be non-null, so it must be called first.
-  // The explicit ctor makes it less likely we accidentally move the owned websocket first.
-  return ItemsForRelease(getWebSocket(lock), kj::mv(KJ_REQUIRE_NONNULL(hibernatableWebSocket.ws)));
+  return ItemsForRelease(kj::mv(websocketRef), kj::mv(ownedWebSocket));
 }
 
-jsg::Ref<WebSocket> HibernatableWebSocketEvent::getWebSocket(jsg::Lock& lock) {
-  auto& actor = KJ_REQUIRE_NONNULL(IoContext::current().getActor());
-  auto& manager = kj::downcast<HibernationManagerImpl>(
-      KJ_REQUIRE_NONNULL(actor.getHibernationManager()));
-  auto& hibernatableWebSocket = KJ_REQUIRE_NONNULL(manager.webSocketForEventHandler);
-  return hibernatableWebSocket.getActiveOrUnhibernate(lock);
+jsg::Ref<WebSocket> HibernatableWebSocketEvent::getWebSocket(jsg::Lock& lock,
+    kj::StringPtr websocketId) {
+  // Should only be called once per event since it removes the HibernatableWebSocket from the
+  // webSocketsForEventHandler collection.
+  auto& manager = kj::downcast<HibernationManagerImpl>(getHibernationManager(lock));
+
+  // Grab it from our collection.
+  auto& hibernatableWebSocket = *KJ_REQUIRE_NONNULL(
+      manager.webSocketsForEventHandler.find(websocketId));
+
+  // Get the reference.
+  auto websocket = hibernatableWebSocket.getActiveOrUnhibernate(lock);
+
+  // Now that we've obtained the websocket, we need to remove the entry from the map and make the
+  // key available again.
+  manager.webSocketsForEventHandler.erase(websocketId);
```

Comment:
I believe if you do `findEntry` above you can erase by reference here. (Which is more or less what happens for `Table::eraseMatch()`.)